### ISSUE TITLE
Yarn install removes everything from bower_components directory

### ISSUE THREAD

**Issue by skuridin** [yarnpkg/yarn#616]
**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**
Removes everything from `bower_components` directory

**If the current behavior is a bug, please provide the steps to reproduce.**
In a project, that uses both npm and bower, run `yarn`

**What is the expected behavior?**
Install missed modules to `node_modules` and do nothing with `bower_components`

**Please mention your node.js, yarn and operating system version.**
macOS 10.12, node 6.6.0, npm 3.10.3, yarn 0.15.1

**Comment by kittens** [comment]
Yarn is actually a drop in replacement for bother Bower and npm. Are we correctly populating `bower_components`? Can you check `node_modules` to see if we're accidentally putting them in there instead?

**Comment by skuridin** [comment]
> Can you check node_modules to see if we're accidentally putting them in there instead?

Couldn't find them in `node_modules`.

`git status` after executing `yarn` (it removes event .gitignore):

```
	deleted:    bower_components/.gitignore
	deleted:    bower_components/jquery-ujs/.bower.json
	deleted:    bower_components/jquery-ujs/CHANGELOG.md
	deleted:    bower_components/jquery-ujs/MIT-LICENSE.txt
	deleted:    bower_components/jquery-ujs/README.md
	deleted:    bower_components/jquery-ujs/bower.json
	deleted:    bower_components/jquery-ujs/src/rails.js
	deleted:    bower_components/jquery.payment/.bower.json
	deleted:    bower_components/jquery.payment/.gitignore
	deleted:    bower_components/jquery.payment/.travis.yml
	deleted:    bower_components/jquery.payment/Cakefile
	deleted:    bower_components/jquery.payment/LICENSE
	deleted:    bower_components/jquery.payment/README.md
	deleted:    bower_components/jquery.payment/bower.json
	deleted:    bower_components/jquery.payment/example/index.html
	deleted:    bower_components/jquery.payment/lib/jquery.payment.js
	deleted:    bower_components/jquery.payment/package.json
	deleted:    bower_components/jquery.payment/payment.jquery.json
	deleted:    bower_components/jquery.payment/src/jquery.payment.coffee
	deleted:    bower_components/jquery.payment/test/index.coffee
```

**Comment by jerguslejko** [comment]
@skuridin do they appear in the listing when you run `yarn ls`?

**Comment by nathanpalmer** [comment]
I'm running into the same issue. Once I run `yarn` and it installs or changes dependencies then `bower_components` is completely empty. 

Looks like they are in `node_modules` instead.

![image](https://cl.ly/1I1V3s223p1X/Screen%20Shot%202016-10-11%20at%2012.43.13%20PM.png)

After I run `bower install` again I get the dependency in bower_components (which is where ember-cli is looking for it.)

![image](https://cl.ly/3Q2V3b0S0B1T/Screen%20Shot%202016-10-11%20at%2012.43.28%20PM.png)

**Comment by RustyToms** [comment]
Same thing here, it appears that bower_components is completely emptied out. I see some of the files in the node_modules folder, but not all of them, I think it might be coincidence that some packages match? Running v0.15.1 on a mac.

If yarn interacts with bower, could that be in the documentation somewhere? I can't even find the word "bower" in the [docs](https://yarnpkg.com/). And it only shows up in any description in the repo [here](https://github.com/yarnpkg/yarn/blob/0f1db96846badbbe3b43c470c05196eb29437fb4/resources/win-chocolatey/yarn.nuspec#L11)

**Comment by skuridin** [comment]
> @skuridin do they appear in the listing when you run yarn ls?

nope

**Comment by RustyToms** [comment]
I made a very simple repo to demonstrate this issue with a single bower.json and a single package.json dependency: https://github.com/RustyToms/yarn-issue-616

**Comment by ethanfrogers** [comment]
I'm also experiencing this behavior but it's removing my `public` directory.

**Comment by RustyToms** [comment]
What seems to be happening here is [bower is being added as a registry](https://github.com/yarnpkg/yarn/blob/master/src/registries/index.js#L10). All the folders in the registries, including `bower_components` are being set as [possibleExtraneous](https://github.com/yarnpkg/yarn/blob/master/src/package-linker.js#L143). Then, if those files weren't included in the files that yarn is building, they are deemed to be actually extraneous and [are deleted](https://github.com/yarnpkg/yarn/blob/master/src/util/fs.js#L103).

The main problem seems to be that bower.json dependencies don't seem to be properly loaded into the set of dependencies that yarn is collecting.

**Comment by MaximRouiller** [comment]
Tried on our project. 

We have a bower.json + package.json and we have the same behavior. We are also using .bowerrc (output to "libs" folder) and this folder also get removed. 

As soon as this is fixed, we can transition on our team.

**Comment by ambirdsall** [comment]
Ran into a fairly weird version of this bug. My project doesn't use bower, but a collaborator used it once to install a dependency. They made a `.bowerrc` file like the following:
```json
{
  "directory": "js/"
}
```
so that it would be installed in the `js` directory with the application code. I had moved the dependency to `node_modules` with the others, but forgot the `.bowerrc` file because it's hidden. Running `yarn` ended up deleting all of my application code; as there was no output indicating `.bowerrc` was being read, it took quite some time to figure a hidden file for a program that isn't part of my current workflow was to blame.

Deleting extraneous module locations makes a certain amount of sense, but doing so silently (a confirmation before delete would be ideal, but at least logging what's happening would have saved me a fair bit of time) is rather user-unfriendly—at least in edge cases like this one.

**Comment by klowdo** [comment]
Why do you include bower_components in your source constrol?

**Comment by olragon** [comment]
I am developing an app based on Drupal 7, all libraries should be saved in `sites/all/libraries` (include PHP library, non bower, non npm library). `yan add something` and everything is gone now :sob:

**Comment by barryvdh** [comment]
Yeah it's pretty annoying, hope everyone trying out Yarn either has the entire folder in version control, or can be rebuild with bower install..

**Comment by Gioni06** [comment]
I can confirm that `yarn add` and `yarn remove` clears the contents of the `bower_components` folder. I'm using a `bower.json` with a `.bowerrc` file

**Comment by kitsunde** [comment]
Since yarn 0.16 it's possible to hook in bower install and make this a bit less distracting while we wait for a proper fix. In package.json:

```json
{
  "scripts": {
    "postinstall": "bower install"
  }
}
```

**Comment by bstro** [comment]
when i `yarn add <packagename>` for anything, it also clears out my bower_components dir

**Comment by agusputra** [comment]
Same issue here. yarn clear bower_components after `yarn add` and `yarn remove`

**Comment by linuxonrails** [comment]
I have the same problem: 
I want to use yarn *and* bower (for other packages) :-(

**Comment by bloodlanser** [comment]
+1. The same issue :(

**Comment by omnisite** [comment]
+1, same issue

**Comment by warapitiya** [comment]
Same issue here.

yarn version 16.1
Ubuntu 16.04

**Comment by kittens** [comment]
We've removed support for Bower, see #1441.

**Comment by linuxonrails** [comment]
Same issue here.

Why should I use only bower or yarn? why not both?!

I have (old) packages in bower_components/ directory and I want to replace npm with yarn (I can use npm with bower).

**Comment by gabrielkoerich** [comment]
You can use yarn and bower. We just have to wait for a new tag.

**Comment by mischah** [comment]
> Why should I use only bower or yarn? why not both?!

That will be possible with the next release of yarn because Yarn won’t take care of your Bower packages. See #1441

Tip. Use `yarn install` and define `bower install` as `postinstall` script in your `package.json`.

**Comment by linuxonrails** [comment]
OK! I'm sorry!!!  :+1: 
Thanks!!!

**Comment by shafi49** [comment]
I didn't expect this kind of behavior. I was excited about yarn. But after wasting half an hour figuring out where my bower_components go I gave up on yarn.

**Comment by Arrow7000** [comment]
Our of curiosity, when will the next, non-bower-components-removing version of Yarn be released? 
Thanks

**Comment by jamesdixon** [comment]
bump

**Comment by sinedied** [comment]
Please publish the version that fixes bower, this is really blocking here :(

**Comment by omnisite** [comment]
Just use the workaround with postinstall. Works like a charm. Anyway: bower support will be dropped, so the workaround is your only option.

**Comment by elmariofredo** [comment]
@omnisite problem is that `yarn add something` will clean bower_components directory as well

**Comment by omnisite** [comment]
Ah ok. Well, you can always just edit the package and bower json file. I know "add" is easier, but there is just not much choice if Yarn is dropping support for Bower.

**Comment by elmariofredo** [comment]
@omnisite you can do all sort of hacks like run `yarn` right after `yarn add` but that's not the point, as yarn is not supporting bower then it should not mess with bower_components folder and prevent a lot of confusion in Polymer and other bower infected communities all around the globe.

Therefore prompt release would be more than appreciated 😉 

Thanks!

**Comment by omnisite** [comment]
@elmariofredo Yes, I agree. I was bummed also that it did not work out of the box. Still I can work with both with the workaround. And we have no idea when they will fix it.

**Comment by elmariofredo** [comment]
@omnisite sure just please mind that OSS is driven by community comments and stating 

> Works like a charm

is IMHO not exactly fitting this case

**Comment by sinedied** [comment]
@omnisite Your workaround does not work in my case, as bower dependencies are commited within the repo (we don't like to rely on the bower tool on each build as we had issues with it from time to time).

yarn deprecating bower support is just fine for me, as long as it does not mess with it.

**Comment by skddc** [comment]
Also, it seems to be done already, so why not release?

https://github.com/yarnpkg/yarn/pulls?utf8=%E2%9C%93&q=is%3Apr%20is%3Aclosed%20bower

**Comment by dennisfrank** [comment]
Yes . Please release!

I ran into this problem myself while traveling offline. yarn’s offline mode was the primary reason to use yarn in the first place. Without a local backup I had not been able to restore bower_components. This is serious. The workarounds from the comments above don’t work offline.

Thanks!

**Comment by patcoll** [comment]
@dennisfrank Thanks for saying that.

I'm tempted to offer a bounty for someone re-instating and completing bower support. It's still widely used, even if it is being deprecated in a lot of places. And its ability to have repeatable installs is very poor.

**Comment by barryvdh** [comment]
I don't think the biggest problem is missing Bower, it's the unexpected deletion of the bower directory :/

Not really sure why this isn't seen as a critical enough issue to release a new (minor) tag for, this can cause serious problems for someone who isn't using version control.

**Comment by dennisfrank** [comment]
@barryvdh Exactly. But even if you you use version control the bower directory often gets ignored on purpose.

**Comment by MaximRouiller** [comment]
Once this get published, I'm ditching npm for our team. Meanwhile? I'll wait,

**Comment by thetrevdev** [comment]
Not sure what is blocking the release or the target eta.  With an early library like this it would be better to see them make changes faster.  This has been sitting for a month.

**Comment by gyujincho** [comment]
I have same issue here, wonder why the fixed version is not being released :(

**Comment by MaximRouiller** [comment]
No idea. In the mid time, this is preventing adoption.

**Comment by elmariofredo** [comment]
It's released with version 0.17.0 🎉

**Comment by cyyyu** [comment]
YES. Version 0.17.0 have fixed this. Already upgraded.

**Comment by shafi49** [comment]
yes... it is true guys. yarn is safe for bower users. thanks yarn team

**Comment by MaximRouiller** [comment]
Updated yarn to latest with:
```
npm install -g yarn
```

Ran `yarn --version` and it resolved to `0.17.0.

Ran `yarn` on my solution and only detected 1 change: yarn.lock

You guys rock. We're putting a change in our backlog to change to yarn today. 😄

**Comment by MaximRouiller** [comment]
![out-npm](https://cloud.githubusercontent.com/assets/209384/20310589/9480a632-ab1a-11e6-96d0-66397f0b6fa7.jpg)

[Original](http://trappedinvacancy.deviantart.com/art/Defenestration-115846260)
